File: minecommons-paper/src/main/java/com/github/aecsocket/minecommons/paper/effect/ParticleEffect.java
package com.github.aecsocket.minecommons.paper.effect;

import com.github.aecsocket.minecommons.core.vector.cartesian.Vector3;

import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * A Paper implementation of a particle effect.
 */
public record ParticleEffect(
        org.bukkit.Particle name,
        int count,
        Vector3 size,
        double speed,
        @Nullable Object data
) implements com.github.aecsocket.minecommons.core.effect.ParticleEffect {
    @Override
    public String toString() {
        return name.name() + " x" + count + " " + size + " @ " + speed +
                (data == null ? "" : "[" + data + "]");
    }
}

File: paper/src/main/kotlin/com/gitlab/aecsocket/alexandria/paper/serializer/ParticleEffect.kt
package com.gitlab.aecsocket.alexandria.paper.serializer

import com.gitlab.aecsocket.alexandria.core.extension.force
import com.gitlab.aecsocket.alexandria.core.physics.Vector3
import com.gitlab.aecsocket.alexandria.paper.effect.ParticleEffect
import org.bukkit.Particle
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.kotlin.extensions.get
import org.spongepowered.configurate.serialize.TypeSerializer
import java.lang.reflect.Type

private const val PARTICLE = "particle"
private const val COUNT = "count"
private const val SIZE = "size"
private const val SPEED = "speed"
private const val DATA = "data"

object ParticleEffectSerializer : TypeSerializer<ParticleEffect> {
    override fun serialize(type: Type, obj: ParticleEffect?, node: ConfigurationNode) {
        if (obj == null) node.set(null)
        else {
            node.node(PARTICLE).set(obj.particle)
            node.node(COUNT).set(obj.count)
            node.node(SIZE).set(obj.size)
            node.node(SPEED).set(obj.speed)
            obj.data?.let { node.node(DATA).set(it) }
        }
    }

    override fun deserialize(type: Type, node: ConfigurationNode): ParticleEffect {
        val particle = node.node(PARTICLE).force<Particle>()
        val data = when (val dataType = particle.dataType) {
            Void::class.java -> null
            /* note: it's valid to have the data be null here
               even if the particle is deserialized with no data,
               and the particle requires data,
               it can always be provided through code later */
            else -> node.node(DATA).get(dataType)
        }
        return ParticleEffect(
            particle,
            node.node(COUNT).get { 0.0 },
            node.node(SIZE).get { Vector3.Zero },
            node.node(SPEED).get { 0.0 },
            data
        )
    }
}
